src/Http/Routing/src/Builder/EndpointFilterExtensions.cs

```diff
@@ -13,12 +13,15 @@ namespace Microsoft.AspNetCore.Http;
 /// </summary>
 public static class EndpointFilterExtensions
 {
+    internal const string FilterRequiresDynamicCodeWarning = "Filter factories generate dynamic code and aren't compatible with native AOT applications.";
+
     /// <summary>
     /// Registers a filter onto the route handler.
     /// </summary>
     /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
     /// <param name="filter">The <see cref="IEndpointFilter"/> to register.</param>
     /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the route handler.</returns>
+    [RequiresDynamicCode(FilterRequiresDynamicCodeWarning)]
     public static TBuilder AddEndpointFilter<TBuilder>(this TBuilder builder, IEndpointFilter filter) where TBuilder : IEndpointConventionBuilder =>
```

Comment:
@captainsafia @davidfowl - Will these filter APIs work correctly in AOT when we have the minimal API source generator?

Reply:
The API to get the value today is something like `T GetValue<T>(...)`, accessed via `MakeGenericMethod`. Won't work with a struct parameter.

Adding a `object GetValue(...)` method might be a way to get AOT compatibility. There would be the cost of boxing, but better than an error. The alternative would be figuring out code gen.

Reply:
How important is avoiding the boxing? Would we be comfortable with using `object GetValue(...)` everywhere irrespective of the runtime? (I.e. box even if we had a JIT.)

If boxing wouldn't be acceptable elsewhere, would it be possible to extract the problematic pattern into a standalone console app so that we can zoom in on it separately and see if something could be done on the runtime side?

Reply:
Yes it will work with the source generator. We have a prototype. This needs to be removed.

Reply:
> This RequiresDynamicCode needs to be removed.

I believe then it means that the suppressions added here:

https://github.com/dotnet/aspnetcore/blob/9175347822be11e4aa1376e69a4c5f17db731388/src/Http/Routing/src/RouteEndpointDataSource.cs#L114-L122

need to be removed as well? Basically meaning even the `MapAction(RequestDelegate)` overloads need to be marked as `RequiresDynamicCode`. And only the source generator will enable AOT'ing minimal APIs successfully.

Reply:
> Basically meaning even the MapAction(RequestDelegate) overloads need to be marked as RequiresDynamicCode. And only the source generator will enable AOT'ing minimal APIs successfully.

No, this code needs to be refactored so that there are abstractions that flow into this code. The source generator will reuse this code with a source generated implementation. The Map(Delegate) calls will add the unsafe implementation and the Map(RequestDelegate) overloads will add a safe implementation.